### PR TITLE
WIP: Attempt to remove the dependencies on fs for browser usage

### PR DESCRIPTION
**Pull request by jdalrymple**
## Summary
fix: Removed xhr library in favour of ky, and switched request for got for a smaller package size and retry functionality #98 

breaking: Removed dependency on FS. Now the Projects API takes in two arguments `projectId` and `content` as well as an option fileName argument

## Testing
1. Attempt to upload a file to a project https://docs.gitlab.com/ee/api/projects.html#upload-a-file through this library
2. Test a few other endpoints to make sure the got library replacement worked
3. Test this library in the browser to see if the replacement through the browser package.json change worked properly using ky instead of got

### CODE REVIEW

**Comment by jdalrymple** [comment]
Still testing! :sunglasses:

**Comment by jdalrymple** [comment]
Not sure if im doing the browser bundling correctly:

```
    "build:browser": "browserify src/index.ts -p [ tsify --noImplicitAny ] > dist/browser.js",
```

but then i get this error:

```
> gitlab@4.2.0 build:browser /home/hyde/Projects/Libraries/node-gitlab
> browserify src/index.ts -p [ tsify --noImplicitAny ] > dist/browser.js

TypeScript error: node_modules/@types/node/index.d.ts(6406,21): Error TS2451: Cannot redeclare block-scoped variable 'promisify'.
```

Any ideas?

**Comment by jdalrymple** [comment]
Finally sorted out the above issue, but ran into another problem. The replacement library Ky, for handling requests has been build to support the latest browsers. I have no issue just supporting the latest browsers for the browser build of this library. 

That being said, it will still have to be bundled. If we use the default Ky library, we cannot use browserify since it will only work with es5 modules. This leaves a couple options:

1. Compile the Ky library down to es5 and use browserify
2. Don't use browserify, instead use another bundler such as webpack

Thoughts?

**Comment by jdalrymple** [comment]
Yup! Switched to rollup and everything seems to be working except https://github.com/TrySound/rollup-plugin-terser/issues/11. Hopefully i get that sorted soon!

File: src/templates/ResourceAwardEmojis.ts [hunk 1]
```diff
@@ -1,15 +1,14 @@
-import URLJoin from 'url-join';
 import { BaseService, RequestHelper } from '../infrastructure';
 
 function url(projectId, resourceType, resourceId, awardId, noteId) {
   const [pId, rId] = [projectId, resourceId].map(encodeURIComponent);
-  let output = `${pId}/${resourceType}/${rId}/`;
+  const output = [pId, resourceType, rId];
```
Comment (casz) [inline src/templates/ResourceAwardEmojis.ts]: I approve! I usually prefer to deal with arrays and just join them later.

File: src/infrastructure/RequestHelper.ts [hunk 2]
```diff
@@ -1,7 +1,5 @@
-import Humps from 'humps';
-import LinkParser from 'parse-link-header';
-import URLJoin from 'url-join';
 import Request from 'got';
+import { decamelizeKeys } from 'humps';
```
Comment (casz) [inline src/infrastructure/RequestHelper.ts]: less KB 🙌

File: src/infrastructure/RequestHelper.ts [hunk 3]
```diff
@@ -11,69 +9,60 @@ interface GetPaginatedOptions {
...
-    if (stream) {
-      return Request.stream(
-        ...defaultRequest(service, endpoint, {
-          query: options,
-        }),
-      );
-    }
-
-    const response = await getPaginated(service, endpoint, options);
+    if (!stream) return getPaginated(service, endpoint, options);
```
Comment (casz) [inline src/infrastructure/RequestHelper.ts]: sometimes negative statements are cleaner

File: src/templates/ResourceMembers.ts [hunk 4]
```diff
@@ -1,19 +1,19 @@
...
-    this.url = URLJoin(this.url, resourceType);
+    this.url = [this.url, resourceType].join('/');
   }
 
   all(resourceId, includeInherited = false, options = {}) {
     const rId = encodeURIComponent(resourceId);
+    const url = [rId, 'members'];
 
-    const url = includeInherited ? `${rId}/members/all` : `${rId}/members`;
+    if (includeInherited) url.push('all');
```
Comment (casz) [inline src/templates/ResourceMembers.ts]: yup definite improvement

File: .gitignore [hunk 5]
```diff
@@ -2,3 +2,4 @@ node_modules
 npm-debug.log
 dist
 coverage
+.rpt2_cache
```
Comment (casz) [inline .gitignore]: hmmm maybe `.editorconfig` would be good to add and set insert final new line. Your making the next edit harder without final new line

File: .prettierrc.yml [hunk 6]
```diff
@@ -0,0 +1,3 @@
+printWidth: 100
+singleQuote: true
+trailingComma: es6
```
Comment (casz) [inline .prettierrc.yml]: This could be moved to package.json 🤭

File: package.json [hunk 7]
```diff
@@ -1,83 +1,97 @@
...
+    "rollup-plugin-json": "^3.1.0",
+    "rollup-plugin-node-resolve": "^3.4.0",
+    "rollup-plugin-terser": "^3.0.0",
+    "rollup-plugin-typescript2": "^0.17.1",
+    "semantic-release": "^15.10.3",
     "ts-jest": "^23.10.4",
     "tslint": "^5.11.0",
     "tslint-config-airbnb": "^5.11.0",
-    "typescript": "^3.1.2"
+    "tslint-config-prettier": "^1.15.0",
```
Comment (casz) [inline package.json]: thank you for adding prettier, would be great with prettier being a devDep too and include a npm script to check/fix formatting.

File: package.json [hunk 8]
```diff
@@ -1,83 +1,97 @@
...
+    "@semantic-release/git": "^7.0.5",
+    "@semantic-release/npm": "^5.0.5",
     "@types/humps": "^1.1.2",
     "@types/jest": "^23.3.5",
     "@types/parse-link-header": "^1.0.0",
     "@types/qs": "^6.5.1",
     "@types/request-promise": "^4.1.42",
     "@types/url-join": "^0.8.2",
     "@types/util.promisify": "^1.0.0",
+    "awesome-typescript-loader": "^5.2.1",
```
Comment (casz) [inline package.json]: Seems like a unused import?

Comment (casz) [inline package.json]: Seems like a unused dep?

File: package.json [hunk 9]
```diff
@@ -1,83 +1,97 @@
...
-    "@semantic-release/npm": "^5.0.4",
+    "@semantic-release/changelog": "^3.0.1",
+    "@semantic-release/git": "^7.0.5",
+    "@semantic-release/npm": "^5.0.5",
     "@types/humps": "^1.1.2",
     "@types/jest": "^23.3.5",
     "@types/parse-link-header": "^1.0.0",
     "@types/qs": "^6.5.1",
     "@types/request-promise": "^4.1.42",
     "@types/url-join": "^0.8.2",
```
Comment (casz) [inline package.json]: Unused dep after refactor

File: package.json [hunk 10]
```diff
@@ -1,83 +1,97 @@
...
-    "@semantic-release/changelog": "^3.0.0",
-    "@semantic-release/git": "^7.0.4",
-    "@semantic-release/npm": "^5.0.4",
+    "@semantic-release/changelog": "^3.0.1",
+    "@semantic-release/git": "^7.0.5",
+    "@semantic-release/npm": "^5.0.5",
     "@types/humps": "^1.1.2",
     "@types/jest": "^23.3.5",
     "@types/parse-link-header": "^1.0.0",
     "@types/qs": "^6.5.1",
```
Comment (casz) [inline package.json]: qs dep was removed 😇

File: package.json [hunk 11]
```diff
@@ -1,83 +1,97 @@
...
   "devDependencies": {
-    "@semantic-release/changelog": "^3.0.0",
-    "@semantic-release/git": "^7.0.4",
-    "@semantic-release/npm": "^5.0.4",
+    "@semantic-release/changelog": "^3.0.1",
+    "@semantic-release/git": "^7.0.5",
+    "@semantic-release/npm": "^5.0.5",
     "@types/humps": "^1.1.2",
     "@types/jest": "^23.3.5",
     "@types/parse-link-header": "^1.0.0",
```
Comment (casz) [inline package.json]: parse-link-header was removed 😇

File: package.json [hunk 12]
```diff
@@ -1,83 +1,97 @@
...
     "@types/qs": "^6.5.1",
     "@types/request-promise": "^4.1.42",
     "@types/url-join": "^0.8.2",
     "@types/util.promisify": "^1.0.0",
```
Comment (casz) [inline package.json]: util-promisify dep was removed

File: package.json [hunk 13]
```diff
@@ -1,83 +1,97 @@
...
-    "@semantic-release/git": "^7.0.4",
-    "@semantic-release/npm": "^5.0.4",
+    "@semantic-release/changelog": "^3.0.1",
+    "@semantic-release/git": "^7.0.5",
+    "@semantic-release/npm": "^5.0.5",
     "@types/humps": "^1.1.2",
     "@types/jest": "^23.3.5",
     "@types/parse-link-header": "^1.0.0",
     "@types/qs": "^6.5.1",
     "@types/request-promise": "^4.1.42",
```
Comment (casz) [inline package.json]: request-promise dep was removed

File: .prettierrc.yml (hunk 6 above)
Comment (jdalrymple) [reply .prettierrc.yml]: I tried to keep as much config out of the package.json :P

File: .gitignore (hunk 5 above)
Comment (jdalrymple) [reply .gitignore]: Any suggestions for a good editorconfig?

Comment (casz) [reply .gitignore]: This seems to apply for the whole repo, of course you can apply different rules for each file type: https://editorconfig.org/
```
root = true

[*]
end_of_line = lf
charset = utf-8
insert_final_newline = true
ident_style = space
ident_size = 2
```

File: .prettierrc.yml (hunk 6 above)
Comment (casz) [inline .prettierrc.yml]: final new line (change was before editorconfig)

**Comment by jdalrymple** [comment]
Just need to figure out the tests then i can merge this!

**Comment by casz** [comment]
Looks really promising 🙌

**Comment by casz** [comment]
Looks really promising 🙌

**Comment by casz** [comment]
Looks really promising 🙌

**Comment by casz** [comment]
Looks really promising 🙌

**Comment by jdalrymple** [comment]
Trying to sort issue with named imports https://github.com/rollup/rollup/issues/2521

**Comment by jdalrymple** [comment]
Hmm even when i disable treeshaking, the es output is still broken. The top two lines of the output file are:

```
import * as Request from 'got';
import { get, stream, post, put, delete } from 'got';
```

I'm not sure 1. why the second line is present since i included the treeshaking: false argument and 2. why both these lines are present to begin with.

Ill keep investigating

**Comment by jdalrymple** [comment]
Found a way around the treeshaking problem ;)

**Comment by jdalrymple** [comment]
https://github.com/sindresorhus/got/pull/657
https://github.com/sindresorhus/ky/pull/61

**Comment by jdalrymple** [comment]
Couldnt get around it, had to include the esModuleInterop

**Pull request merged**